### PR TITLE
[TDF] Custom actions: Fix Book and provide tutorial showing how to fill ThN!

### PR DESCRIPTION
**Pull request by dpiparo**

### CODE REVIEW

File: tutorials/dataframe/tdf018_customActions.C [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+   using Result_t = THn_t;
+
+private:
+   std::vector<std::shared_ptr<THn_t>> fHistos; // one per slot
+   const ROOT::Detail::TDF::ColumnNames_t fColumnNames;
+
+public:
+   THnHelper(std::string_view name, std::string_view title, std::array<int, NDIM> nbins, std::array<double, NDIM> xmins,
+             std::array<double, NDIM> xmax, unsigned int nSlots, ROOT::Detail::TDF::ColumnNames_t columnNames)
+      : fHistos(nSlots, std::make_shared<THn_t>(std::string(name).c_str(), std::string(title).c_str(), NDIM,
```
Comment (bluehood) [inline tutorials/dataframe/tdf018_customActions.C]: this creates a vector of shared_ptrs that all point to the same address (the one allocated by the make_shared call):

```c++
#include <cassert>
#include <memory>
#include <vector>

int main() {
   std::vector<std::shared_ptr<int>> ptrs(2, std::make_shared<int>(42));
   assert(ptrs[0] != ptrs[1]);
   return 0;
}
```

File: tutorials/dataframe/tdf018_customActions.C [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+   }
+   THnHelper(THnHelper &&) = default;
+   THnHelper(const THnHelper &) = delete;
+   ROOT::Detail::TDF::ColumnNames_t GetColumnNames() const { return fColumnNames; }
+   std::shared_ptr<THn_t> GetResultPtr() const { return fHistos[0]; }
+   void Initialize() {}
+   void InitTask(TTreeReader *, unsigned int) {}
+   void Exec(unsigned int slot, ColumnTypes... values)
+   {
+      std::array<double, sizeof...(ColumnTypes)> valuesArr{(double)values...};
```
Comment (bluehood) [inline tutorials/dataframe/tdf018_customActions.C]: `static_cast<double>(values)...`? :grin:

File: tutorials/dataframe/tdf018_customActions.C [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      fHistos[slot]->Fill(valuesArr.data());
+   }
+
+   void Finalize()
+   {
+      auto &res = fHistos[0];
+      for (auto slot : ROOT::TSeqU(1, fHistos.size())) {
+         res->Add(fHistos[slot].get());
```
Comment (bluehood) [inline tutorials/dataframe/tdf018_customActions.C]: is this strictly the same as `Merge`?

File: tutorials/dataframe/tdf018_customActions.C [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      }
+   }
+};
+
+void tdf018_customActions()
+{
+   const auto nSlots = 4;
+   ROOT::EnableImplicitMT(nSlots);
```
Comment (bluehood) [inline tutorials/dataframe/tdf018_customActions.C]: this being a tutorial maybe we shouldn't show that we force the number of slots to a particular value

File: tutorials/dataframe/tdf018_customActions.C (hunk 1 above)
Comment (dpiparo) [reply tutorials/dataframe/tdf018_customActions.C]: ahaha, what a mistake! nice catch.

File: tutorials/dataframe/tdf018_customActions.C (hunk 2 above)
Comment (dpiparo) [reply tutorials/dataframe/tdf018_customActions.C]: ... fine ...

File: tutorials/dataframe/tdf018_customActions.C (hunk 3 above)
Comment (dpiparo) [reply tutorials/dataframe/tdf018_customActions.C]: Yes, but without TList.

File: tutorials/dataframe/tdf018_customActions.C (hunk 4 above)
Comment (dpiparo) [reply tutorials/dataframe/tdf018_customActions.C]: Yes. On the other hand we don't want to overcommit the infra when running the tests even though it's not going to be a very heavy workload...

**Comment by dpiparo** [comment]
Given that we approach the freeze, I will merge it tomorrow in order to be able to keep an eye on incrementals.

**Comment by bluehood** [comment]
In the helper's ctor it seems there is an extra `nSlots` parameter to `make_shared`, and the loop variable `i` will cause "unused variable" warnings. Besides these it looks good to me.

**Pull request merged**